includes/utilities.inc

```diff
@@ -582,3 +583,73 @@ function islandora_multi_importer_isValidXmlName($dsid) {
   }
 }
 
+/**
+ * Constructs a URI to IMI's file storage location for a file.
+ * Create the directory if it does not exist.
+ *
+ * @param string $filename
+ *   The name of the file to be stored.
+ *   Required.
+ * @param string $imi_temp_subdir
+ *   A sub-path.
+ *   Optional.
+ * @param string $imi_temp_dir
+ *   The islandora-multi-importer directory location relative to the uri scheme.
+ *   Optional. Defaults to 'islandora-multi-importer'.
+ * @param string $scheme
+ *   The name of the drupal stream wrapper scheme.
+ *   Optional. Defaults to "private".
+ *
+ * @return string
+ *   Returns a normalized stream wrapper uri to the file storage location.
+ *   Returns FALSE if unable to create/verify the file storage directory.
+ */
+function _imi_build_file_storage_uri($filename, $imi_temp_subdir = NULL, $imi_temp_dir = NULL, $scheme = 'private' ) {
+
+  $directory_uri = _imi_build_file_storage_directory($imi_temp_subdir, $imi_temp_dir, $scheme);
+  if(empty($directory_uri)) {
+    drupal_set_message(t('Error occurred while preparing storage for @filename in @directory. Check destination folder permissions', array(
+      '@filename' => $filename,
+      '@directory' => $directory_uri,
+    )), 'error');
+    return FALSE;
+  }
+  // Assemble and return the full file uri.
+  return file_stream_wrapper_uri_normalize($directory_uri . DIRECTORY_SEPARATOR . $filename);
+}
+
+/**
+ * Creates a temporary file storage directory
+ *
+ * @param string   $imi_temp_subdir
+ *   A subdirectory path.
+ *   Optional - None will be used if not provided.
+ * @param string   $imi_temp_dir
+ *   The main imi temporary files directory.
+ *   Optional - Default will be used if none is provided.
+ * @param string $scheme
+ *   The Drupal filestream scheme (public://, private://, temporary://).
+ *   Optional - private:// will be used if none is provided.
+ *
+ * @return bool/string
+ *   The uri to the new directory if successful.
+ *   FALSE if unable to create the directory.
+ */
+function _imi_build_file_storage_directory($imi_temp_subdir = NULL, $imi_temp_dir = NULL, $scheme = 'private') {
+
+  $directory = array(!empty($imi_temp_dir) ? $imi_temp_dir : variable_get('islandora_multi_importer_temp_dir', 'islandora-multi-importer'));
+
+  if(!empty($imi_temp_subdir)) {
+    $directory[] = $imi_temp_subdir;
+  }
+  $directory_uri = file_stream_wrapper_uri_normalize($scheme . '://' . implode(DIRECTORY_SEPARATOR, $directory));
+
+  $writable = file_prepare_directory($directory_uri, FILE_CREATE_DIRECTORY);
+  if(!$writable) {
+    drupal_set_message(t('Error occurred while preparing destination directory @directory. Check permissions', array(
```

Comment:
We need a message, or hint, that allows people to fix the issue if they encounter this problem. Are we checking if directory is writeable on the admin page first?

Reply:
We could. My thinking is that I didn't want to get into ensuring writability prior to actually trying to write. We cannot assume that the IMI settings page has been visited and saved. And even if we had, we still have to check for writability at the time we try to write the file, since that's the only time we know for sure whether the file can be written. 
If you think this is important, I'll go ahead and add a form validate function that checks to see if the directory can be created.

Reply:
Yes, please, i agree we need to keep that check everytime people update this, but validating also the form could help too